working/static metaprogramming/intro.md

```diff
@@ -0,0 +1,421 @@
+   method declaration, etc. We haven’t spent time here yet.
+2. Metaprogramming happens by defining a macro, which can be called at compile
+   time.
+3. The introspection system is TBD. It could be an imperative API similar to the
+   existing mirrors API or analyzer API (but run at compile-time), or perhaps
+   something declarative like to pattern matching over syntax examples.
+4. We could use an imperative API to construct new pieces of Dart code in a
+   macro, but some sort of quotation syntax would make this nicer and make the
+   API less brittle.
+
+Most of this is as-yet-undesigned, but the key piece is that **we think macros
+should be written in normal imperative Dart code which is executed at compile
+time**.
+
+## Compile-time Dart execution
+
+Allowing a general-purpose, Turing-complete imperative language to run at
+compile time is a big ask. Here’s why we think it’s a good choice:
+
+### We already do it
+
+A subset of the Dart language is already allowed at compile time, which is
+specifically carved out to support limited const expressions. This means users
+are familiar with the concept of “running” Dart code at compile time, and we
+have invested some implementation time in supporting it. You could think of this
+proposal as merely (massively) expanding the subset of Dart allowed in a const
+expression.
+
+### One less language to define
+
+Metaprogramming has to be authored in some language. It takes significant time
+to design, specify, implement, and test a language. Anyone user authoring macros
+will have to learn that language. By making that Dart language, we let the user
+write macros in a language they already know, and we take advantage of all of
+he existing work we’ve done on Dart.
+
+### Declarative languages are too limited
+
+Declarative macro systems are usually not Turing complete. This ensures that
+metaprogramming has finite execution time, which is a nice property. The cost is
+that many potentially useful metaprogramming abstractions can’t be defined
+within the limitations of that declarative language. Experience from other
+language communities is that declarative-only metaprogramming ends up being too
+limited.
+
+The Scheme community has gone through a number of rules-based macro systems over
+decades as each proved to be too limiting. Meanwhile Common Lisp let Lispers
+define macros imperatively. This sometimes makes it a challenge to implement
+macros *correctly*, but has always been sufficiently powerful.
+
+Experience from many UI templating languages (which are surprisingly similar to
+macro metaprogramming in many ways) is that most declarative ones gradually
+accumulate imperative features or end up replaced by those that do.
+
+In the Dart ecosystem, in the absence of metaprogramming in the language, users
+often resort to code generators. Some of those are quite sophisticated and are
+written in imperative Dart code.
+
```

Comment:
might be worth calling out Dart kernel transformers here which perhaps have a superset of the functionality you need.

Reply:
done